### PR TITLE
sslyze 0.14.0

### PR DESCRIPTION
**Pull request by ilovezfs**
use new virtualenv feature
build nassl from source instead of using upstream prebuilt binaries
add head spec
run test suite

### CODE REVIEW

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@tdsmith it seems that pip is allowing nassl to install a prebuilt binary ... that seems, um, bad.

**Comment by tdsmith** [comment]
Can you point me to what makes you think that's happening?

**Comment by ilovezfs** [comment]
build log: https://gist.github.com/ilovezfs/b065eb614f83a486a00508bb745d541e

```
    clang -bundle -undefined dynamic_lookup -arch i386 -arch x86_64 -Wl,-F. build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_SSL_CTX.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_SSL.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_X509.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_errors.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_BIO.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_X509_EXTENSION.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_X509_NAME_ENTRY.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_SSL_SESSION.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/openssl_utils.o build/temp.macosx-10.11-intel-2.7/nassl/_nassl/nassl_OCSP_RESPONSE.o /private/tmp/pip-LxyOzh-build/bin/openssl/darwin64/libssl.a /private/tmp/pip-LxyOzh-build/bin/openssl/darwin64/libcrypto.a /private/tmp/pip-LxyOzh-build/bin/zlib/darwin64/libz.a -o build/lib.macosx-10.11-intel-2.7/nassl/_nassl.so
```

tar file contents:
```
Josephs-MacBook-Pro:~ joe$ tar -tf <(curl -Ls https://files.pythonhosted.org/packages/40/f0/f787c59b65210b04bc23d6c459db31e32b8bb117baec2b4bafc5120e7f2d/nassl-0.13.7.tar.gz)
nassl-0.13.7/
nassl-0.13.7/bin/
nassl-0.13.7/bin/openssl/
nassl-0.13.7/bin/openssl/darwin64/
nassl-0.13.7/bin/openssl/darwin64/libcrypto.a
nassl-0.13.7/bin/openssl/darwin64/libssl.a
nassl-0.13.7/bin/openssl/include/
nassl-0.13.7/bin/openssl/include/openssl/
nassl-0.13.7/bin/openssl/include/openssl/aes.h
nassl-0.13.7/bin/openssl/include/openssl/asn1.h
nassl-0.13.7/bin/openssl/include/openssl/asn1_mac.h
nassl-0.13.7/bin/openssl/include/openssl/asn1t.h
nassl-0.13.7/bin/openssl/include/openssl/bio.h
nassl-0.13.7/bin/openssl/include/openssl/blowfish.h
nassl-0.13.7/bin/openssl/include/openssl/bn.h
nassl-0.13.7/bin/openssl/include/openssl/buffer.h
nassl-0.13.7/bin/openssl/include/openssl/camellia.h
nassl-0.13.7/bin/openssl/include/openssl/cast.h
nassl-0.13.7/bin/openssl/include/openssl/chacha.h
nassl-0.13.7/bin/openssl/include/openssl/cmac.h
nassl-0.13.7/bin/openssl/include/openssl/cms.h
nassl-0.13.7/bin/openssl/include/openssl/comp.h
nassl-0.13.7/bin/openssl/include/openssl/conf.h
nassl-0.13.7/bin/openssl/include/openssl/conf_api.h
nassl-0.13.7/bin/openssl/include/openssl/crypto.h
nassl-0.13.7/bin/openssl/include/openssl/des.h
nassl-0.13.7/bin/openssl/include/openssl/des_old.h
nassl-0.13.7/bin/openssl/include/openssl/dh.h
nassl-0.13.7/bin/openssl/include/openssl/dsa.h
nassl-0.13.7/bin/openssl/include/openssl/dso.h
nassl-0.13.7/bin/openssl/include/openssl/dtls1.h
nassl-0.13.7/bin/openssl/include/openssl/e_os2.h
nassl-0.13.7/bin/openssl/include/openssl/ebcdic.h
nassl-0.13.7/bin/openssl/include/openssl/ec.h
nassl-0.13.7/bin/openssl/include/openssl/ecdh.h
nassl-0.13.7/bin/openssl/include/openssl/ecdsa.h
nassl-0.13.7/bin/openssl/include/openssl/engine.h
nassl-0.13.7/bin/openssl/include/openssl/err.h
nassl-0.13.7/bin/openssl/include/openssl/evp.h
nassl-0.13.7/bin/openssl/include/openssl/hmac.h
nassl-0.13.7/bin/openssl/include/openssl/idea.h
nassl-0.13.7/bin/openssl/include/openssl/krb5_asn.h
nassl-0.13.7/bin/openssl/include/openssl/kssl.h
nassl-0.13.7/bin/openssl/include/openssl/lhash.h
nassl-0.13.7/bin/openssl/include/openssl/md2.h
nassl-0.13.7/bin/openssl/include/openssl/md4.h
nassl-0.13.7/bin/openssl/include/openssl/md5.h
nassl-0.13.7/bin/openssl/include/openssl/mdc2.h
nassl-0.13.7/bin/openssl/include/openssl/modes.h
nassl-0.13.7/bin/openssl/include/openssl/obj_mac.h
nassl-0.13.7/bin/openssl/include/openssl/objects.h
nassl-0.13.7/bin/openssl/include/openssl/ocsp.h
nassl-0.13.7/bin/openssl/include/openssl/opensslconf.h
nassl-0.13.7/bin/openssl/include/openssl/opensslv.h
nassl-0.13.7/bin/openssl/include/openssl/ossl_typ.h
nassl-0.13.7/bin/openssl/include/openssl/pem.h
nassl-0.13.7/bin/openssl/include/openssl/pem2.h
nassl-0.13.7/bin/openssl/include/openssl/pkcs12.h
nassl-0.13.7/bin/openssl/include/openssl/pkcs7.h
nassl-0.13.7/bin/openssl/include/openssl/poly1305.h
nassl-0.13.7/bin/openssl/include/openssl/pqueue.h
nassl-0.13.7/bin/openssl/include/openssl/rand.h
nassl-0.13.7/bin/openssl/include/openssl/rc2.h
nassl-0.13.7/bin/openssl/include/openssl/rc4.h
nassl-0.13.7/bin/openssl/include/openssl/rc5.h
nassl-0.13.7/bin/openssl/include/openssl/ripemd.h
nassl-0.13.7/bin/openssl/include/openssl/rsa.h
nassl-0.13.7/bin/openssl/include/openssl/safestack.h
nassl-0.13.7/bin/openssl/include/openssl/seed.h
nassl-0.13.7/bin/openssl/include/openssl/sha.h
nassl-0.13.7/bin/openssl/include/openssl/srp.h
nassl-0.13.7/bin/openssl/include/openssl/srtp.h
nassl-0.13.7/bin/openssl/include/openssl/ssl.h
nassl-0.13.7/bin/openssl/include/openssl/ssl2.h
nassl-0.13.7/bin/openssl/include/openssl/ssl23.h
nassl-0.13.7/bin/openssl/include/openssl/ssl3.h
nassl-0.13.7/bin/openssl/include/openssl/stack.h
nassl-0.13.7/bin/openssl/include/openssl/symhacks.h
nassl-0.13.7/bin/openssl/include/openssl/tls1.h
nassl-0.13.7/bin/openssl/include/openssl/ts.h
nassl-0.13.7/bin/openssl/include/openssl/txt_db.h
nassl-0.13.7/bin/openssl/include/openssl/ui.h
nassl-0.13.7/bin/openssl/include/openssl/ui_compat.h
nassl-0.13.7/bin/openssl/include/openssl/whrlpool.h
nassl-0.13.7/bin/openssl/include/openssl/x509.h
nassl-0.13.7/bin/openssl/include/openssl/x509_vfy.h
nassl-0.13.7/bin/openssl/include/openssl/x509v3.h
nassl-0.13.7/bin/openssl/include/openssl-internal/
nassl-0.13.7/bin/openssl/include/openssl-internal/e_os.h
nassl-0.13.7/bin/openssl/include/openssl-internal/ssl_locl.h
nassl-0.13.7/bin/openssl/linux32/
nassl-0.13.7/bin/openssl/linux32/libcrypto.a
nassl-0.13.7/bin/openssl/linux32/libssl.a
nassl-0.13.7/bin/openssl/linux64/
nassl-0.13.7/bin/openssl/linux64/libcrypto.a
nassl-0.13.7/bin/openssl/linux64/libssl.a
nassl-0.13.7/bin/openssl/openbsd64/
nassl-0.13.7/bin/openssl/openbsd64/libcrypto.a
nassl-0.13.7/bin/openssl/openbsd64/libssl.a
nassl-0.13.7/bin/openssl/win32/
nassl-0.13.7/bin/openssl/win32/libeay32.lib
nassl-0.13.7/bin/openssl/win32/ssleay32.lib
nassl-0.13.7/bin/openssl/win64/
nassl-0.13.7/bin/openssl/win64/libeay32.lib
nassl-0.13.7/bin/openssl/win64/ssleay32.lib
nassl-0.13.7/bin/zlib/
nassl-0.13.7/bin/zlib/darwin64/
nassl-0.13.7/bin/zlib/darwin64/libz.a
nassl-0.13.7/bin/zlib/linux32/
nassl-0.13.7/bin/zlib/linux32/libz.a
nassl-0.13.7/bin/zlib/linux64/
nassl-0.13.7/bin/zlib/linux64/libz.a
nassl-0.13.7/bin/zlib/openbsd64/
nassl-0.13.7/bin/zlib/openbsd64/libz.a
nassl-0.13.7/bin/zlib/win32/
nassl-0.13.7/bin/zlib/win32/zlibstat.lib
nassl-0.13.7/bin/zlib/win64/
nassl-0.13.7/bin/zlib/win64/zlibstat.lib
nassl-0.13.7/LICENSE.txt
nassl-0.13.7/nassl/
nassl-0.13.7/nassl/__init__.py
nassl-0.13.7/nassl/_nassl/
nassl-0.13.7/nassl/_nassl/nassl.c
nassl-0.13.7/nassl/_nassl/nassl_BIO.c
nassl-0.13.7/nassl/_nassl/nassl_BIO.h
nassl-0.13.7/nassl/_nassl/nassl_errors.c
nassl-0.13.7/nassl/_nassl/nassl_errors.h
nassl-0.13.7/nassl/_nassl/nassl_OCSP_RESPONSE.c
nassl-0.13.7/nassl/_nassl/nassl_OCSP_RESPONSE.h
nassl-0.13.7/nassl/_nassl/nassl_SSL.c
nassl-0.13.7/nassl/_nassl/nassl_SSL.h
nassl-0.13.7/nassl/_nassl/nassl_SSL_CTX.c
nassl-0.13.7/nassl/_nassl/nassl_SSL_CTX.h
nassl-0.13.7/nassl/_nassl/nassl_SSL_SESSION.c
nassl-0.13.7/nassl/_nassl/nassl_SSL_SESSION.h
nassl-0.13.7/nassl/_nassl/nassl_X509.c
nassl-0.13.7/nassl/_nassl/nassl_X509.h
nassl-0.13.7/nassl/_nassl/nassl_X509_EXTENSION.c
nassl-0.13.7/nassl/_nassl/nassl_X509_EXTENSION.h
nassl-0.13.7/nassl/_nassl/nassl_X509_NAME_ENTRY.c
nassl-0.13.7/nassl/_nassl/nassl_X509_NAME_ENTRY.h
nassl-0.13.7/nassl/_nassl/openssl_utils.c
nassl-0.13.7/nassl/_nassl/openssl_utils.h
nassl-0.13.7/nassl/debug_ssl_client.py
nassl-0.13.7/nassl/ocsp_response.py
nassl-0.13.7/nassl/ssl_client.py
nassl-0.13.7/nassl/x509_certificate.py
nassl-0.13.7/PKG-INFO
nassl-0.13.7/README.md
nassl-0.13.7/setup.cfg
nassl-0.13.7/setup.py
```

It's using prebuilt
```
nassl-0.13.7/bin/openssl/darwin64/libcrypto.a
nassl-0.13.7/bin/openssl/darwin64/libssl.a
```

**Comment by DomT4** [comment]
The custom SSL that was in this formula is semi-legit because of the need to test SSLv2/SSLv3 ciphers which most sensible people have started taking an axe to. Not sure if the resource containing the binary here is related to that.

**Comment by ilovezfs** [comment]
I think there's a case this should be boneyarded because it has a hard dependency on nassl, whose github homepage says

>Experimental OpenSSL wrapper for Python 2.7 and SSLyze. Do NOT use for anything serious. This code has not been properly tested/reviewed and is absolutely not production ready.

That doesn't exactly sound like something we'd accept if this were a new formula.

**Comment by ilovezfs** [comment]
@DomT4 more generally, what's very concerning here is that even though we pass ``"--no-binary", ":all:"` pip apparently is still totally happy to install prebuilt binaries, which means we have basically no idea if any of the python formulae are pip installing random binary rubbish from pypi tarballs.

**Comment by DomT4** [comment]
I presume `--no-binary` only means Python wheels, eggs, etc rather than embedded non-python binaries in source packages, but Tim will know more. Less than 40 downloads, so if you wanted to kill it, no strong objection.

**Comment by ilovezfs** [comment]
@DomT4 yeah pretty scary.

**Comment by ilovezfs** [comment]
@DomT4 PR refreshed.

File: Formula/sslyze.rb [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+    sha256 "5aa0e159cd0c118b6ae99ec12095707e5a7bd53c13dceeadd153ce767d941b66"
+
+    # pypi version has prebuilt binaries
+    resource "nassl" do
+      url "https://github.com/nabla-c0d3/nassl/archive/0.13.7.tar.gz"
+      sha256 "98512ac55d38d0839f795248f321ad9bf6971a3c0fb7224e84ad4f5c81a5bbcb"
+    end
+
+    resource "openssl" do
+      url "https://github.com/PeterMosmans/openssl.git",
```
Comment (DomT4) [inline Formula/sslyze.rb]: Why the unofficial fork of OpenSSL?

File: Formula/sslyze.rb [hunk 2]
```diff
@@ -12,53 +27,62 @@ class Sslyze < Formula
...
+    end
   end
 
+  depends_on :arch => :x86_64
+  depends_on :python if MacOS.version <= :snow_leopard
+
   resource "zlib" do
-    url "http://zlib.net/zlib-1.2.8.tar.gz"
-    sha256 "36658cb768a54c1d4dec43c3116c27ed893e88b02ecfcb44f2166f9c0b7f2a0d"
+    url "http://http.debian.net/debian/pool/main/z/zlib/zlib_1.2.8.dfsg.orig.tar.gz"
```
Comment (DomT4) [inline Formula/sslyze.rb]: Bleh. I keep meaning to audit this to death. Let's stick to the secure Debian mirrors 🙈.

File: Formula/sslyze.rb (hunk 1 above)
Comment (ilovezfs) [inline Formula/sslyze.rb]: Upstream's choice. https://github.com/PeterMosmans/openssl see the readme. The idea is to enable insecure ciphers and such for maximal coverage for testing/analysis as opposed to a library you actually *use* for your own production systems.

Comment (DomT4) [inline Formula/sslyze.rb]: Yeah, I'm aware of Peter's fork; I shipped it myself in a tap for a while to play with it. I hope Peter is aware other projects are officially starting to require his fork 😅.

Comment (ilovezfs) [inline Formula/sslyze.rb]: https://github.com/nabla-c0d3/nassl

>If you do not want to use the pre-compiled binaries, compiling the C extension requires successively building:
>
>    Zlib 1.2.8
>    A special fork of OpenSSL 1.0.2 (or the official OpenSSL 1.0.2e)
>    The _nassl C extension itself

not sure why `e` lol

Comment (ilovezfs) [inline Formula/sslyze.rb]: probably the last to have enough old ciphers to not break the build or something ...

Comment (DomT4) [inline Formula/sslyze.rb]: OpenSSL turned off SSLv2 by default on one of the 1.0.2 releases. Possibly 1.0.2e from that message.

It broke API compatibility for MacPorts, I remember, and they had to do a partial rebuild. Made me thankful we'd done it yonks ago.

Comment (ilovezfs) [inline Formula/sslyze.rb]: What's MacPorts? :trollface:

**Comment by tdsmith** [comment]
what Dom said.

> which means we have basically no idea if any of the python formulae are pip installing random binary rubbish from pypi tarballs.

This is, like, true of any tarball we install. :p I promise they're not; exceptions that prove the rule etc

**Comment by tdsmith** [comment]
Whoops, mashed the wrong button, sorry

**Comment by DomT4** [comment]
What's the status on this?

**Comment by ilovezfs** [comment]
@DomT4 status is ready-to-merge-when-:green_apple:

**Comment by ilovezfs** [comment]
:ship: 'd

**Pull request closed without merging**